### DISCUSSION TITLE
feature request: Threads & Other Channel Types

### DISCUSSION THREAD

**Discussion by Ombucha** [stoatchat/discussions#476] [Feature Suggestions]
### What do you want to see?

- Threads will allow users to discuss multiple topics in the same channel.
- Other channel types such as **Stage Channels**, **Forum Channels**, **News Channels**, **Rules Channels**, etc. will improve the app further.

### PWA

- [ ] Yes, this feature request is specific to the PWA.

**Comment by JeroenAdam** [comment]
I've not seen it done yet, but the ability to tag threads with a custom label would be awesome, that way links can be created between threads (within the same discussion).
Let's imagine the Revoltchat Project tracker as a giant discussion, each card on the project tracker represented by a thread, well, some threads could be tagged 'enhancement', others 'bug', etc.

**Comment by do-me** [comment]
### Zulip-style threads

I'd just like to add my 2 cents here: 
We are currently using Zulip at work and it's great due to the innovative threading model - couldn't live without it anymore. 

However, I tried creating an online community on [gis.chat](https://gis.chat) where Zulip is not (yet) well-suited. The feedback was that it's **not** fun to use and it's sort of true. I was considering Revolt as a replacement as it has a much nicer UI/UX but it's lacking a way to structure conversations better. 

What I really hate about all Discord/Slack/Telegram groups is the lack of structure. If you're absent for a while you just can't understand what messages to read and which to skip so you end up wasting much time skimming through all messages instead of applying some kind of filter.

Have a look at this [fork of Zulip with hashtags](https://github.com/v3ss0n/zulip/pull/2) and the respective [discussion](https://github.com/zulip/zulip/issues/13928) that could maybe serve as inspiration to this discussion and hopefully the development of Revolt.

**Reply by JeroenAdam** [reply]
I can second that, in theory you (and I) could use the server abstraction in Revolt as a channel and the channel abstraction as a thread.
Let's look at a url in Revolt: https://app.host/server/xxxxxxxxx/channel/yyyyyyyyyy
Let's now look at a url in RocketChat: https://app.host/group/xxxxxxxxx/thread/yyyyyyyyyy
It would mean that such an improvised channel would be accessible as a server via the round icons most left in the sidebar.
Revolt additionally has the notion of categories allowing channels (aka threads in this example), to be ordered in the sidebar.
I understood that servers are just abstractions running inside one Revolt instance.

**Reply by do-me** [reply]
@JeroenAdam I know what you mean - it helps imagining the "levels of nesting".

I think there is no need to re-invent the wheel as Zulip did some great work in thinking this through. Consider the public instance for example. There are tools to see/filter all messages on the server.

![image](https://github.com/revoltchat/revolt/assets/47481567/bf7822e7-5c72-4265-995e-b53764096731)

If you click on any of the streams you get all messages of all topics in that stream. If you click on one, you'll see only the messages in that topic. 

![image](https://github.com/revoltchat/revolt/assets/47481567/672478f0-79e4-49b4-8d11-23ae9ae7a39e)

Streams are ordered alphabetically atm (but you can filter them which is handy).

I think both platforms can learn a lot from each other.

**Comment by eatallyourdarlings** [comment]
**Boost!** I've been looking for a D*scord alternative to help form an creative online community, and the lack of this feature in Revolt means that I'll probably have to keep looking for options, or learn javascript to help contribute...

Where threads really shine in other apps is where they allow you to have mini-channels and let people go off track without clogging up a channel. My use-case would be having a channel for creatives to chat about 'project ideas' and if two people really get chatting one of them can opt to turn the conversation into a thread and name it. This way you can let these conversations fizzle out and search for them later to pick them back up again. It also makes catching up on channels a lot easier. Element.io has a good feature where in addition to just replying to messages in a standard way, you can 'reply in thread,' including replying to yourself to start a thread.

**Reply by do-me** [reply]
This is already possible with Zulip by the way. 
You would write in a channel's "general" topic (=thread) until there's the need to create a dedicated one for the conversation. Then you just move the messages to this new topic. The messages can still be seen chronologically in the channel (if you see all ongoing topics) but the new topic gives you the threads functionality. Have a look at the [public instance](https://chat.zulip.org) to see it in action. 
This feature is also one of the reasons I opted for Zulip in the first place but it really encourages (if not forces) you to structure everything.

**Reply by eatallyourdarlings** [reply]
Yeah I had a look at Zulip, and it's too much for my use case. A comfortable middle ground can definitely be found with Zulip's functionality, like a cleaner version of Discord's threads implementation.

**Reply by do-me** [reply]
Yep, I understand - I guess this middle ground is also what I'm looking for but it's hard to give up on some features once you know them. :D
Element threads for me are ok-ish but once you have +20 ongoing threads it becomes unmanageable (no naming, no filtering etc.). Unfortunately, I think there's more work tbd before they are actually useful for large group channels.

**Comment by D1reW0lf** [comment]
Frankly if Revolt ever get threads and some other options it might as well do as Guilded does instead of what Discord tried to do. 
Even though Discord is late to the *thread party* their implementation is way worst than the one Guild already has since ever.

**Comment by adamjrice** [comment]
+1 to this. I think forum channels would be more useful than threading, simply because people are inconsistent about replying in threads, but forum channels essentially force you into a thread. Also, for my selfish purposes, forum channels would be a useful addition.

**Comment by AllBubbledUp** [comment]
+1 to all of these, especially the idea for rules channels.

i think it'd be cool to have a channel that

- is read-only and starts at top by default
- has no channel header ("This is the start of your conversation.")
- and has no visible message sender

similar to a server guide resource on discord, but shown prominently on the channel list instead of tucked away into a tab.

**Comment by 25Adam25** [comment]
+1 to this.
would be lovely!

**Comment by negativefnnancy** [comment]
to speak kind of practically , for me i wld sort of like some kind of thread-like system that essentially is tailored to the following situations:
- conversation branching (eg going off on a temporary tangent unrelated to main thread of conversation, to prevent clogging)
- "transient discussion channels" , ie situation where it makes sense to have a dedicated channel, but its only temporarily relevant, something u wldnt want to be sitting there in the main channel list taking up space bc its not really a "main channel" (eg main channel for showcasing projects, and subchannel for further discussion pertaining to that posting)
- "movable subchannels" , sometimes i actually find myself deciding that a channel i created, which has already accumulated discussion, is more of "transient nature" , and i may want to reclassify it as a "thread"/"subchannel" , and mayb i just want to be able to drag it in the channel list "into" another main channel so that it is hierarchically a child channel (but this hierarchy wld be strictly limited to a depth of 1) , or vice verse: a subthread was made and accumulates discussion, but i decide it shld be reclassified as a "main channel" and want to "drag it out" into the main channel list , and like id sorta rather avoid having to have a tucked away "channel archival category" for a lot of such things

i think i essentially feel a want for a functional distinction between "fixed directory/main channels" (which id want to comprise the servers always-visible top level channel listing) and "transient/context specific channels" (which want their own functional channel as a discussion partition, but also dont want to occupy space in the servers main channel directory)

basically a need to keep both discussion itself sufficiently partitioned in a way spontaneously adaptive to actual discussion
as well as to keep the main/fixed server channel directory listing orderly

**Comment by rogue-agent** [comment]
Missing threads feature and possibility to make public discussion

Proposal
Having a threads type system for a channel type where users of Stoat and users of threadiverse can interact per topic.
<img width="1453" height="725" alt="stoatforum" src="https://github.com/user-attachments/assets/b717c777-baa0-456d-a731-3a2b89cd9110" />

Image depicting a list view of some threads in a forum type threaded channel. Image also contains users from outside the Stoat network who have posted, either connecting their account to Stoat or the server broadcasting the forum space publicly. Image also uses Stoat logos for samples as to where an image would be located if the topic contained an image.
